### PR TITLE
Feature/electrical meter

### PR DESCRIPTION
**Pull request by aleemont1**
This PR introduces a basic Electrical Meter.

### CODE REVIEW

**Comment by aleemont1** [comment]
This is a very basic implementation. It doesn't take count of the amount of time an outlet has been consuming power. It cuts power to the most-consuming outlet in the very moment the global consumption is equal to the global max consumption value. 

In the next updates I want to implement the following features:

- Checking the average consumption in the last n. ticks of the simulation (of a single outlet and of all the outlets, too).
- Cutting the power if the average consumption of the last n. ticks is above the max power.

But still idk how to do so, I'll study the clock class already implemented but I'll listen for advice.

**Comment by smagnaniuni** [comment]
> This is a very basic implementation. It doesn't take count of the amount of time an outlet has been consuming power. It cuts power to the most-consuming outlet in the very moment the global consumption is equal to the global max consumption value.
> 
> In the next updates I want to implement the following features:
> 
>     * Checking the average consumption in the last n. ticks of the simulation (of a single outlet and of all the outlets, too).
> 
>     * Cutting the power if the average consumption of the last n. ticks is above the max power.
> 
> 
> But still idk how to do so, I'll study the clock class already implemented but I'll listen for advice.
> 
> For a more advanced level i think I might need to check wether the derivative of the average consumption in the last tick is positive (The consumption is rising) or negative (The consumption is decreasing), but I'll see.
> 
> Maybe it should be done from the main controller, in the main loop, I'll think of it.

For simulation matters see #49 
Each device could implement a SimElement interface with an updateTick(dt) / updateStep(dt) / however we want to call it.
In there the deltas/steps/derivatives can be multiplied by the time difference and get the actual value out.
The controller would also implement it.
All of the updateTicks would be called from the simulation side (our "game" loop)

**Comment by aleemont1** [comment]
I think it will need refactoring due to last changes to interfaces, I'll check and do it asap.

**Comment by aleemont1** [comment]
Please give me a feedback otherwise I'm stuck with this.

**Comment by michelececcacci** [comment]
didn't you say you needed to do a refactor?

File: app/src/main/java/homer/controller/api/electricalmeter/ElectricalMeter.java
Comment (michelececcacci) [inline app/src/main/java/homer/controller/api/electricalmeter/ElectricalMeter.java:51]: not sure we need something like this, honestly.  I don't see the point of having this public

File: app/src/main/java/homer/controller/impl/electricalmeter/ElectricalMeterImpl.java
Comment (michelececcacci) [inline app/src/main/java/homer/controller/impl/electricalmeter/ElectricalMeterImpl.java:58]: i think this can be done in a nicer way with streams

File: app/src/main/java/homer/controller/impl/electricalmeter/ElectricalMeterImpl.java
Comment (michelececcacci) [inline app/src/main/java/homer/controller/impl/electricalmeter/ElectricalMeterImpl.java:48]: shouldn't this be marked with a `throws` clause?

**Review by michelececcacci — CHANGES_REQUESTED** [review]

**Comment by aleemont1** [comment]
I noticed you changed the directory structure, is it necessary for me to do the same in this pr? Should I just create the new dir structure or merge from dev?

**Comment by michelececcacci** [comment]
As i said before, `computeconsumption` can be more readable by  using streams.

**Comment by michelececcacci** [comment]
>I noticed you changed the directory structure, is it necessary for me to do the same in this pr? Should I just create the new dir structure or merge from dev?

you need to merge, otherwise we can't track changes

**Review by michelececcacci — CHANGES_REQUESTED** [review]
upstream changes need to be merged

**Comment by aleemont1** [comment]
Done. Should be ok now. I'll wait for your review.

**Comment by aleemont1** [comment]
> For simulation matters see #49 Each device could implement a SimElement interface with an updateTick(dt) / updateStep(dt) / however we want to call it. In there the deltas/steps/derivatives can be multiplied by the time difference and get the actual value out. The controller would also implement it. All of the updateTicks would be called from the simulation side (our "game" loop)

So ElectricalMeter should extend DiscreteObject, and Override the updateTick method, in which I'd basically call "checkConsumption", right?

**Comment by michelececcacci** [comment]
Should be good. please fix CI errors.

**Comment by aleemont1** [comment]
> Should be good. please fix CI errors and assign `"COUTLET"` and `"LOUTLET"` to a variable each in the unit tests

Should I assign constants for every number I used in tests? That seems a bit odd. Can we write a post on the forum to ask if somebody has solved this linter issue? Or maybe just to know if it's ok to suppress this warning on test classes.

**Comment by aleemont1** [comment]
@michelececcacci 
https://github.com/checkstyle/checkstyle/issues/2584

**Comment by michelececcacci** [comment]
should be good. Please verify that checkstyle passes

**Comment by michelececcacci** [comment]
LGTM

**Comment by michelececcacci** [comment]
please resolve merge conflicts before asking for reviews

**Comment by aleemont1** [comment]
DON'T MERGE BEFORE MERGING Outlets

**Comment by aleemont1** [comment]
> 

Done. I think you can resolve this. IDKW it looks CI checks haven't been running since that change request.

**Comment by aleemont1** [comment]
> upstream changes need to be merged

Done. Looks like CI checks haven't been running since this change request, and I can't resolve the conversation. Can you do it?

**Comment by michelececcacci** [comment]
still merge conflicts

**Comment by aleemont1** [comment]
> still merge conflicts

Yep, probably due to last feature/Outlets merge, I'll fix it asap.

**Comment by aleemont1** [comment]
CI is now passing, I'm merging this.

**Pull request merged**
